components/zcash_address/src/kind/unified.rs

```diff
@@ -0,0 +1,141 @@
+            _ => Ok(Receiver::Unknown {
+                typecode,
+                data: addr.to_vec(),
+            }),
+        }
+        .map_err(|_| ParseError::InvalidEncoding)
+    }
+}
+
+impl Receiver {
+    fn typecode(&self) -> u8 {
+        match self {
+            Receiver::P2pkh(_) => 0x00,
+            Receiver::P2sh(_) => 0x01,
+            Receiver::Sapling(_) => 0x02,
+            Receiver::Orchard(_) => 0x03,
+            Receiver::Unknown { typecode, .. } => *typecode,
+        }
+    }
+
+    fn addr(&self) -> &[u8] {
+        match self {
+            Receiver::P2pkh(data) => data,
+            Receiver::P2sh(data) => data,
+            Receiver::Sapling(data) => data,
+            Receiver::Orchard(data) => data,
+            Receiver::Unknown { data, .. } => data,
+        }
+    }
+}
+
+/// A Unified Address.
+#[derive(Clone, Debug, PartialEq, Eq, Hash)]
+pub struct Address(pub(crate) Vec<Receiver>);
+
+impl TryFrom<&[u8]> for Address {
+    type Error = ParseError;
+
+    fn try_from(buf: &[u8]) -> Result<Self, Self::Error> {
+        let encoded = f4jumble::f4jumble_inv(buf).ok_or(ParseError::InvalidEncoding)?;
+
+        // Validate and strip trailing zero bytes.
+        let encoded = match encoded.split_at(encoded.len() - PADDING_LEN) {
+            (encoded, tail) if tail == &[0; PADDING_LEN][..] => Ok(encoded),
+            _ => Err(ParseError::InvalidEncoding),
+        }?;
+
+        iter::repeat(())
+            .scan(encoded, |encoded, _| match encoded {
+                // Base case: we've parsed the full encoding.
+                [] => None,
+                // The raw encoding of a Unified Address is a concatenation of:
+                // - typecode: byte
+                // - length: byte
+                // - addr: byte[length]
+                [typecode, length, data @ ..] if data.len() >= *length as usize => {
+                    let (addr, rest) = data.split_at(*length as usize);
+                    *encoded = rest;
+                    Some((*typecode, addr).try_into())
+                }
+                // The encoding is truncated.
+                _ => Some(Err(ParseError::InvalidEncoding)),
+            })
+            .collect::<Result<_, _>>()
+            .map(Address)
+    }
+}
+
+impl Address {
+    /// Returns the raw encoding of this Unified Address.
+    pub(crate) fn to_bytes(&self) -> Vec<u8> {
+        self.0
+            .iter()
+            .flat_map(|receiver| {
+                let addr = receiver.addr();
+                // Holds by construction.
+                assert!(addr.len() < 256);
+
+                let encoded: Vec<_> = iter::empty()
+                    .chain(Some(receiver.typecode()))
+                    .chain(Some(addr.len() as u8))
+                    .chain(addr.into_iter().cloned())
+                    .chain(iter::repeat(0).take(PADDING_LEN))
+                    .collect();
+
+                f4jumble::f4jumble(&encoded).unwrap()
```

Comment:
Blocking: I thought f4jumble was intended to be applied to the concatenated bytes (the entire unified address), not to each receiver?

Reply:
You are correct. The parser is also correct, so currently this isn't round-trip compatible. It happens to be passing tests though because we only have a test for a single-receiver address, for which this happens to work.

Reply:
Remember that the input to F4Jumble also includes the 16 zero bytes.

Reply:
Yep, that was also a bug that I fixed at the same time.